### PR TITLE
[FIX] l10n_ch: QR-Invoice don't work for LI customers

### PR DESCRIPTION
**Pull request by momegahed**
Issue :
- Liechtenstein adapted the same QR-Invoice as Switzerland. However, Odoo only allows the issuance of QR-Invoices to swiss customers

https://www.llb.li/en/private/paying-and-saving/payment-services/qr-bill#:~:text=Standing%20orders%20based%20on%20orange%20payment%20slips%20can%20no%20longer%20be%20processed%20after%2030%20September%202022.%20Therefore%2C%20these%20standing%20orders%20need%20to%20be%20newly%20set%20up%20on%20the%20basis%20of%20QR%20bills.

Fix:
- Allow LI users to be issued QR-Invoices

OPW-2977644

### CODE REVIEW

**Comment by jco-odoo** [comment]
@momegahed I see other instances of "CH" in this file and there are probably elsewhere as well.  Is there a way to do a little bit of a broader check?

**Comment by jco-odoo** [comment]
@casp-odoo

**Comment by momegahed** [comment]
@jco-odoo @casp-odoo 

The scope of this PR is to allow a CH company to make QR-invoice for LI customers.

The rest of the checks including `_is_qr_iban` are meant to verify that the company has all the right configurations to issue QR-invoices

However, since now technically LI companies can also issue QR-invoice, the other checks should also be changed to accommodate this albeit in another PR (this raises questions about the module name and relevance since it is the swiss localization)


TLDR;
other checks out of scope of this bug/ticket

**Comment by momegahed** [comment]
@jco-odoo @casp-odoo wdyt ?

**Comment by casp-odoo** [comment]
This allows for a CH user to issue a QR invoice to a LI client, but not the other way around, is that correct? 
I guess changes in stable versions should be kept minimal, so why not, but I also think there's a lack here that should be fulfilled at some point. I'll raise that question.

**Comment by momegahed** [comment]
> This allows for a CH user to issue a QR invoice to a LI client, but not the other way around, is that correct?
this is correct

> I guess changes in stable versions should be kept minimal, so why not, but I also think there's a lack here that should be fulfilled at some point. I'll raise that question.
I agree. it should be added in another module (LI customization)

**Comment by jco-odoo** [comment]
@momegahed I am not sure that the _eligible_for_qr_code is not used also to check that the own company is from Liechtenstein or not.  partner_bank_id on the customer invoice might be the company one. 

It still makes sense to allow the Liechtenstein as a customer, but then we need to see that we do not go into trouble when the company is and we support it half/half.  

What I hear from people who lived in Liechtenstein is that they actually use the same CoA, ... as in Switzerland and I guess you pay the same VAT as well and need to pay domestic 
VAT for both LI/CH customers, so it can make sense to make l10n_ch work for both LI/CH I guess, but then maybe in a later version where the fiscal country could still be CH?  @yvaucher

**Comment by momegahed** [comment]
@yvaucher 👀

**Comment by momegahed** [comment]
Hi @yvaucher, a super gentle reminder

**Comment by yvaucher** [comment]
@momegahed thanks for the reminder. I checked with @jco-odoo during the Odoo XP

:+1: for this PR which solves generation of invoice from CH company to LI customer.

(thanks for the pings)


:information_source:  On a more general note, you will want to consider Liechtenstein as a Switzerland region as it uses the same norms and specs.
So almost anytime you have a `if ... = "CH"` you will want to also make the same condition for `"LI"`. BTW the QR-Bill specs clearly mention LI. (For instance you can use LI IBAN)

**Review by yvaucher — APPROVED** [review]

**Comment by momegahed** [comment]
> @momegahed I had the change to check this with @jco-odoo during the Odoo XP
> 
> 👍 for this PR which solves generation of invoice from CH company to LI customer.
> 
> (thanks for the pings)
> 
> ℹ️ On a more general note, you will want to consider Liechtenstein as a Switzerland region as it uses the same norms and specs. So almost anytime you have a `if ... = "CH"` you will want to also make the same condition for `"LI"`. BTW the QR-Bill specs clearly mention LI. (For instance you can use LI IBAN)

if it could be done in the same l10n module so why not.

**Comment by jco-odoo** [comment]
@casp-odoo

File: addons/l10n_ch/__init__.py [hunk 1]
```diff
@@ -14,8 +14,8 @@ def load_translations(env):
 def init_settings(env):
     '''If the company is localized in Switzerland, activate the cash rounding by default.
     '''
-    # The cash rounding is activated by default only if the company is localized in Switzerland.
-    for company in env['res.company'].search([('partner_id.country_id.code', '=', "CH")]):
+    # The cash rounding is activated by default only if the company is localized in Switzerland or Liechtenstein.
```
Comment (casp-odoo) [inline addons/l10n_ch/__init__.py:17]: I'm not sure about this. Is it really related to the QR Bills generation? Is the legislation the same on this specific point?

File: addons/l10n_ch/models/account_bank_statement.py [hunk 2]
```diff
@@ -9,7 +9,7 @@ class AccountBankStatementLine(models.Model):
     _inherit = "account.bank.statement.line"
 
     def _find_or_create_bank_account(self):
-        if self.company_id.country_id.code == 'CH' and _is_l10n_ch_postal(self.account_number):
+        if self.company_id.country_id.code in ('CH', 'LI') and _is_l10n_ch_postal(self.account_number):
```
Comment (casp-odoo) [inline addons/l10n_ch/models/account_bank_statement.py:12]: will `_is_l10n_ch_postal` pass with a LI address?

**Review by casp-odoo — COMMENTED** [review]
I wrote a couple questions which I have about LI, mostly to make sure we won't come across surprises down the road. 

I also think adding a test including a LI example would prevent forgetting LI on the side for further changes, don't you think ?

File: addons/l10n_ch/__init__.py (hunk 1 above)
Comment (momegahed) [reply addons/l10n_ch/__init__.py:17]: taking this by the word:

> @momegahed I had the change to check this with @jco-odoo during the Odoo XP

👍 for this PR which solves generation of invoice from CH company to LI customer.

(thanks for the pings)

ℹ️ On a more general note, you will want to consider Liechtenstein as a Switzerland region as it uses the same norms and specs.
So almost anytime you have a if ... = "CH" you will want to also make the same condition for "LI". BTW the QR-Bill specs clearly mention LI. (For instance you can use LI IBAN)

I understand that they share exactly the same accounting standards.

File: addons/l10n_ch/models/account_bank_statement.py (hunk 2 above)
Comment (momegahed) [reply addons/l10n_ch/models/account_bank_statement.py:12]: I did a bit of research now , apparently postfinance (the issuer of postal accounts) also provides services in LI: 
I quote:

> PostFinance offers comprehensive payment transaction solutions for banks and financial service providers based in Switzerland or **Liechtenstein**.

https://www.postfinance.ch/en/business/needs/financial-institutions/switzerland.html

I might be wrong though, can you please confirm? @yvaucher

**Comment by jco-odoo** [comment]
Maybe the only concern is that it is in stable and we should see that we do not break anything for partners.  @yvaucher maybe you know someone.

File: addons/l10n_ch/models/account_bank_statement.py (hunk 2 above)
Comment (yvaucher) [reply addons/l10n_ch/models/account_bank_statement.py:12]: @casp-odoo `_is_l10n_ch_postal` is deprecated, as we only use IBAN since 1st October 2022 for all bank transactions, it is not necessary to consider it.

**Comment by yvaucher** [comment]
@jco-odoo I'm ok with those changes, this shouldn't break anything considering currently Swiss <-> Liechtenstein bills were not possible. As this is not a constraint but lift one blocker it's all good to me.

**Comment by apf-c2c** [comment]
LGTM, Works as expected :+1:     
image

File: addons/l10n_ch/models/account_bank_statement.py (hunk 2 above)
Comment (yvaucher) [reply addons/l10n_ch/models/account_bank_statement.py:12]: Thus, even for postfinance accounts, the account owner will get an IBAN instead of postal number.

**Comment by yvaucher** [comment]
Tested with @apf-c2c and works like a charm

**Comment by momegahed** [comment]
Hi @jco-odoo , Can we r+ this?

**Comment by jco-odoo** [comment]
@momegahed We also have the qr_iban part in account_sepa...

**Comment by jco-odoo** [comment]
@robodoo r+
